Sources/TuistCache/Mappers/CacheGraphMutator.swift

```diff
@@ -6,317 +6,151 @@ import TuistSupport
-        precompiledArtifacts: [GraphTarget: AbsolutePath],
-        sources: Set<String>,
-        sourceTargets: inout Set<GraphTarget>,
-        visitedPrecompiledArtifactPaths: inout [GraphTarget: VisitedArtifact?],
-        loadedPrecompiledNodes: inout [AbsolutePath: GraphDependency]
-    ) throws {
-        sourceTargets.formUnion([target])
-        let targetDependency: GraphDependency = .target(name: target.target.name, path: target.path)
-        graphDependencies[targetDependency] = try mapDependencies(
-            graph.dependencies[targetDependency, default: Set()],
-            graph: graph,
-            graphDependencies: &graphDependencies,
+        var newGraph = oldGraph
+        newGraph.dependencies = try mapDependencies(
+            replaceableTargets: replaceableTargets,
             precompiledArtifacts: precompiledArtifacts,
-            sources: sources,
-            sourceTargets: &sourceTargets,
-            visitedPrecompiledArtifactPaths: &visitedPrecompiledArtifactPaths,
-            loadedPrecompiledArtifacts: &loadedPrecompiledNodes
+            graphTraverser: graphTraverser
+        )
+        newGraph.targets = mapTargets(
+            replaceableTargets: replaceableTargets,
+            graphTraverser: graphTraverser
         )
+        return newGraph
     }
 
-    // swiftlint:disable:next function_body_length
-    fileprivate func mapDependencies(
-        _ dependencies: Set<GraphDependency>,
-        graph: Graph,
-        graphDependencies: inout [GraphDependency: Set<GraphDependency>],
+    fileprivate func makeReplaceableTargets(
         precompiledArtifacts: [GraphTarget: AbsolutePath],
         sources: Set<String>,
-        sourceTargets: inout Set<GraphTarget>,
-        visitedPrecompiledArtifactPaths: inout [GraphTarget: VisitedArtifact?],
-        loadedPrecompiledArtifacts: inout [AbsolutePath: GraphDependency]
-    ) throws -> Set<GraphDependency> {
-        var newDependencies: Set<GraphDependency> = Set()
-        try dependencies.forEach { dependency in
-            let graphTraverser = GraphTraverser(graph: graph)
-            let targetDependency: GraphTarget
-            switch dependency {
-            case let .target(name: name, path: path):
-                guard
-                    let target = graphTraverser.target(path: path, name: name)
-                else { return }
-                targetDependency = target
-            // If the dependency is not a target node we keep it.
-            default:
-                newDependencies.insert(dependency)
-                return
+        graphTraverser: GraphTraversing
+    ) throws -> Set<GraphTarget> {
+        let allTargets = graphTraverser.allTargets()
+        let sortedCacheableTargets = try topologicalSort(
+            Array(allTargets),
+            successors: {
+                Array(graphTraverser.directTargetDependencies(path: $0.path, name: $0.target.name))
             }
```

Comment:
I think we have this piece of code in like 5 places around tuist, maybe worth it to make it an extension of Array<Targets>?